src/main/java/bio/terra/common/FutureUtils.java

```diff
@@ -31,56 +31,52 @@ public static <T> List<T> waitFor(final List<Future<T>> futures) {
    * @return The resolved values of the futures. There is no guarantee that the order of returned
    *     values matches the order of the passed in futures
    */
-  public static <T> List<T> waitFor(
-      final List<Future<T>> futures, final Optional<Duration> maxThreadWait) {
+  public static <T> List<T> waitFor(final List<Future<T>> futures, final Duration maxThreadWait) {
 
-    final AtomicReference<Optional<ApiException>> foundFailure =
-        new AtomicReference<>(Optional.empty());
+    final AtomicReference<ErrorReportException> foundFailure = new AtomicReference<>();
     try (Stream<Future<T>> stream = futures.stream()) {
       List<T> returnList =
           stream
               .map(
                   f -> {
                     try {
                       // If a failure was found, all subsequent tasks should be canceled
-                      if (foundFailure.get().isPresent()) {
+                      if (foundFailure.get() != null) {
                         f.cancel(true);
-                      } else if (maxThreadWait.isPresent()) {
-                        return f.get(maxThreadWait.get().toMillis(), TimeUnit.MILLISECONDS);
+                      } else if (maxThreadWait != null) {
+                        return f.get(maxThreadWait.toMillis(), TimeUnit.MILLISECONDS);
                       } else {
                         return f.get();
                       }
                     } catch (TimeoutException e) {
-                      if (!foundFailure.get().isPresent()) {
-                        foundFailure.set(Optional.of(new ApiException("Thread timed out", e)));
-                        // Cancel the thread
-                        f.cancel(true);
-                      }
+                      foundFailure.compareAndSet(
+                          null, new ServiceUnavailableException("Thread timed out", e));
+                      f.cancel(true); // Do we need this?
```

Comment:
I don't actually think we need this (and ditto the cancellation I added below when catching an `InterruptException`) but I wasn't sure.  If we're in these catch blocks, hasn't the future completed exceptionally?  Can someone more future-savvy help me understand?

Reply:
@nmalfroy -- I'd be interested in what you think about these!

Reply:
I guess I was thinking that if any subthreads were left running beyond throwing a timeout exception it might be safer to explicitly call `cancel` (certainly doesn't hurt).  Like if the thread throws a timeout, it could theoretically still have a nested future running I think

Reply:
Sounds good! I added comments to that effect.